### PR TITLE
Add test for out-of-bounds Tensor access on GPU

### LINKED ISSUE #38944: Test asserts in cuda kernels
So far, we have been avoiding testing cuda kernels in the situations where they will generate device assert (e.g. indexing kernels getting invalid indices from user). This is because device assert invalidates cuda context, and no subsequent tests can be run in the same process. But what is not tested breaks, and we should not leave important behavior untested (see #38943).
One option is to have a separate test suite that would test assertion behavior by launching a process per test, and initializing cuda context in each process. It is extremely slow unfortunately (cuda context initialization takes a few seconds).  

cc @ngimel @mruberry @malfet 
cc @ngimel @mruberry

### PR DESCRIPTION
**Pull request by malfet**
Since CUDA context can not recover safely from on-device assert, use `torch.multiprocessing.spawn` to execute a method in another context and verify that it raises unrecoverable error.

As those types of tests are pretty slow (6 seconds on powerful linux box with one GPU) run it only in the slow shard.

Closes https://github.com/pytorch/pytorch/issues/38944

### CODE REVIEW

**Comment by ngimel** [comment]
You don't have to split the test between the 2 files. Turns out we already have a test like this in test_cuda.py https://github.com/pytorch/pytorch/blob/0e8c65f756c5c24620c843d70f4ceca6a5faeb24/test/test_cuda.py#L1701-L1734, this pattern seems fine to me.

**Comment by ezyang** [comment]
If you don't want to use multiprocessing but still don't want to split across two files you can also do:

```
    @unittest.skipIf(IS_WINDOWS, 'not applicable to Windows (only fails with fork)')
    @unittest.skipIf(not torch.cuda.is_available(), 'CUDA not available')
    def test_wrong_cuda_fork(self):
        results = self.run_process_no_exception("""\
import torch
from torch.multiprocessing import Process
def run(rank):
    torch.cuda.set_device(rank)
if __name__ == "__main__":
    size = 2
    processes = []
    for rank in range(size):
        # it would work fine without the line below
        x = torch.rand(20, 2).cuda()
        p = Process(target=run, args=(rank,))
        p.start()
        processes.append(p)
    for p in processes:
        p.join()
""")
        self.assertRegex(results[1].decode('ascii'), "Cannot re-initialize CUDA in forked subprocess.")
```

File: test/test_cuda.py [hunk 1]
```diff
@@ -2838,6 +2838,20 @@ def test_max_large_axis(self):
...
+        test_path = os.path.join(curdir, 'bounds_test', 'cuda_bounds_test.py')
+
+        # Test in-bounds access succeeds
+        p = subprocess.run([sys.executable, test_path, '1'], capture_output=True, timeout=15)
+        self.assertEqual(p.returncode, 0, msg=f'run of {test_path} failed:{p.stdout}')
+        # Test out-of-bounds access fails
+        p = subprocess.run([sys.executable, test_path, '11'], capture_output=True, timeout=15)
+        self.assertNotEqual(p.returncode, 0, msg=f'run of {test_path} succeeded:{p.stdout}')
+        self.assertIn('cudaErrorAssert', p.stdout.decode('ascii'))
+        self.assertIn('"index out of bounds"', p.stderr.decode('ascii'))
```
Comment (walterddr) [inline test/test_cuda.py:2853]: very good idea to test process level failures. :-)

**Review by ezyang — APPROVED** [review]

File: test/test_cuda.py [hunk 2]
```diff
@@ -1576,6 +1576,38 @@ def test_multinomial_invalid_probs_cuda(self):
...
+            return err
+
+    @slowTest
+    @unittest.skipIf(NO_MULTIPROCESSING_SPAWN, "Disabled for environments that \
+                     don't support multiprocessing with spawn start method")
+    @skipIfRocm
+    def test_index_out_of_bounds_exception_cuda(self):
+        test_method = TestCuda._test_index_bounds_cuda
+        # Test in-bound access works fine
+        self.assertEqual(test_method(1), "x[torch.tensor([1)]=tensor([1], device='cuda:0')")
```
Comment (ngimel) [inline test/test_cuda.py:1607]: this seems too specific?

**Review by ngimel — APPROVED** [review]
Yeah, subprocess part looks good, I'm not sue if we need the "fine" test

**Pull request closed without merging**